src/libraries/Microsoft.PowerFx.Interpreter/PatchFunction.cs

```diff
@@ -0,0 +1,187 @@
+
+            var arg0 = args[0] as RecordValue;
+
+            for (var i = 0; i < args.Length; i++)
+            {
+                foreach (var field in ((RecordValue)args[i]).Fields)
+                {
+                    fieldsDict[field.Name] = field.Value;
+                }
+            }
+
+            var fieldList = new List<NamedValue>();
+
+            foreach (var field in fieldsDict)
+            {
+                fieldList.Add(new NamedValue(field.Key, field.Value));
+            }
+
+            var record = RecordValue.NewRecordFromFields(fieldList);
+
+            return Task.FromResult<FormulaValue>(record);
+        }
+
+        public override RequiredDataSourcePermissions FunctionPermission => RequiredDataSourcePermissions.Create | RequiredDataSourcePermissions.Update;
+    }
+
+    internal class PatchFunction : PatchAndValidateRecordFunctionBase, IAsyncTexlFunction
+    {
+        public override bool IsSelfContained => false;
+
+        public PatchFunction()
+            : base("Patch", TexlStrings.AboutPatch, FunctionCategories.Table | FunctionCategories.Behavior, DType.EmptyRecord, 0, 3, int.MaxValue, DType.EmptyTable, DType.EmptyRecord, DType.EmptyRecord)
+        {
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures()
+        {
+            yield return new[] { TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg };
+            yield return new[] { TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg };
+            yield return new[] { TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg, TexlStrings.PatchChangeRecordsArg };
+        }
+
+        public override IEnumerable<TexlStrings.StringGetter[]> GetSignatures(int arity)
+        {
+            if (arity > 3)
+            {
+                return GetGenericSignatures(arity, TexlStrings.PatchDataSourceArg, TexlStrings.PatchBaseRecordArg, TexlStrings.PatchChangeRecordsArg);
+            }
+
+            return base.GetSignatures(arity);
+        }
+
+        public async Task<FormulaValue> InvokeAsync(FormulaValue[] args, CancellationToken cancel)
+        {
+            var argFields = new Dictionary<string, FormulaValue>();
+
+            var arg0 = args[0] as TableValue;
+            var arg1 = args[1] as RecordValue;
+
+            // Starts from the 3rd argument (new values)
+            for (var i = 2; i < args.Length; i++)
+            {
+                var arg = args[i];
+
+                if (arg is BlankValue)
+                {
+                    continue;
+                }
+                else if (arg is ErrorValue)
+                {
+                    // @@@JYL How to handle error value arguments?
+                    return arg;
+                }
+                else if (arg is RecordValue record)
+                {
+                    // @@@JYL What is a ellegant way to avoid the nested loop?
```

Comment:
> // @@@JYL What is a ellegant way to avoid the nested loop? [](http://example.com/codeflow?start=19&length=59)

nested loop is ok.  It could be extracted into its own function.

Although I wonder - can this code (153...174) be shared with the other overload of patch? via some shared helper

Reply:
> Although I wonder - can this code (153...174) be shared with the other overload of patch? via some shared helper

I believe we can hold for now. The upcoming implementations of Patch don't have the same signature and we are also planning to move the implementation to the interpreter. Let me know if you agree @MikeStall .

Reply:
It looks like this can be shared immediately with the code at line 92.